packages/ember-glimmer/lib/environment.js

```diff
@@ -79,14 +79,20 @@ export default class Environment extends GlimmerEnvironment {
     return false;
   }
 
-  getComponentDefinition(name) {
+  getComponentDefinition(nameSegments, meta) {
+    let name = nameSegments.join('/');
     let definition = this._components[name];
 
     if (!definition) {
-      let { component: ComponentClass, layout } = lookupComponent(this.owner, name[0]);
+      let { component: ComponentClass, layout } = lookupComponent(this.owner, name);
 
       if (ComponentClass || layout) {
-        definition = this._components[name] = new CurlyComponentDefinition(name, ComponentClass, layout);
+        definition = this._components[name] = new CurlyComponentDefinition(nameSegments, ComponentClass, layout);
+      } else if (meta && meta.moduleName) {
+        let segments = meta.moduleName.split('/')
+        segments.shift();
+        segments.push(name);
+        return this.getComponentDefinition(segments);
```

Comment:
moduleName will be somethings like `compoment/parent-foo` and `nameSegments` will be `['my-foo']`. This code creates `segments`, which ends up being `['parent-foo', 'my-foo']` and becomes the `nameSegments` of the recursive call.